File: services/rust/axum/src/routes/handlers/patch.rs
use crate::routes::{
    errors::AppError,
    extractors::{AppJson, AppPath},
};
use app::{state::AppState, task::PatchedTask};
use axum::{
    extract::{Json, State},
    response::{IntoResponse, Response},
};
use std::sync::{Arc, Mutex};
use tracing::{error, info, instrument, warn};

#[axum::debug_handler]
#[instrument(skip_all)]
pub async fn patch_handler(
    State(state): State<Arc<Mutex<AppState>>>,
    AppPath(id): AppPath,
    AppJson(patched_task): AppJson<PatchedTask>,
) -> Response {
    if let Ok(mut state) = state.lock() {
        if let Some(task) = state.tasks.get_mut(&id) {
            let previous_operation = task.operation.clone();
            task.operation = patched_task.operation.clone();

            info!(
                %id,
                secret = task.secret.len(),
                from_operation = previous_operation.to_string(),
                to_operation = task.operation.to_string(),
                method = "PATCH",
                "Patched task"
            );

            return Json(task).into_response();
        } else {
            drop(state);
            warn!(%id, method = "PATCH", "Task not found");
            return AppError::TaskNotFound.into_response();
        }
    }

    error!(%id, method = "PATCH", "Poisoned lock");

    AppError::Internal.into_response()
}

File: services/go/chi/routes/handlers/patch.go
package handlers

import (
	"app"
	"chi/routes"
	"encoding/json"
	"log/slog"
	"net/http"
	"strings"
)

func PatchHandler(state *app.AppState) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id, err := ParseID(w, r)
		if err != nil {
			return
		}

		body, err := ReadBody(w, r)
		if err != nil {
			return
		}

		patchedTask, err := ParsePatchedTask(w, r, body)
		if err != nil {
			return
		}

		state.Mu.Lock()
		task, ok := state.Tasks[id]

		if !ok {
			state.Mu.Unlock()
			slog.Warn("Task not found", "method", r.Method, "path", r.URL.Path, "id", id.String())
			routes.AppErrors.TaskNotFound.Error(w)
			return
		}

		previousOperation := task.Operation
		task.Operation = patchedTask.Operation
		state.Tasks[id] = task
		state.Mu.Unlock()

		slog.Info(
			"Patched task",
			"id", id.String(),
			"secret", len(task.Secret),
			"from_operation", strings.ToLower(string(previousOperation)),
			"to_operation", strings.ToLower(string(patchedTask.Operation)),
			"method", r.Method,
		)

		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		_ = json.NewEncoder(w).Encode(task)
	}
}
